src/main/java/io/stargate/sgv2/jsonapi/api/model/command/CommandContext.java

```diff
@@ -1,86 +1,212 @@
+   * @param commandName
+   * @param jsonProcessingMetricsReporter
+   * @return
+   */
+  public static CommandContext<KeyspaceSchemaObject> forSchemaObject(
+      KeyspaceSchemaObject schemaObject,
+      EmbeddingProvider embeddingProvider,
+      String commandName,
+      JsonProcessingMetricsReporter jsonProcessingMetricsReporter) {
+    return new CommandContext<>(
+        schemaObject, embeddingProvider, commandName, jsonProcessingMetricsReporter);
+  }
 
   /**
-   * @return Returns empty command context, having both {@link #namespace} and {@link #collection}
-   *     as <code>null</code>.
+   * Factory method to create a new instance of {@link CommandContext} based on the schema object we
+   * are working with
+   *
+   * @param schemaObject
+   * @param embeddingProvider
+   * @param commandName
+   * @param jsonProcessingMetricsReporter
+   * @return
    */
-  public static CommandContext empty() {
-    return EMPTY;
+  public static CommandContext<DatabaseSchemaObject> forSchemaObject(
+      DatabaseSchemaObject schemaObject,
+      EmbeddingProvider embeddingProvider,
+      String commandName,
+      JsonProcessingMetricsReporter jsonProcessingMetricsReporter) {
+    return new CommandContext<>(
+        schemaObject, embeddingProvider, commandName, jsonProcessingMetricsReporter);
   }
 
-  public CollectionSettings.SimilarityFunction similarityFunction() {
-    return collectionSettings.vectorConfig().similarityFunction();
+  @SuppressWarnings("unchecked")
+  public CommandContext<CollectionSchemaObject> asCollectionContext() {
+    checkSchemaObjectType(CollectionSchemaObject.TYPE);
+    return (CommandContext<CollectionSchemaObject>) this;
   }
 
-  public boolean isVectorEnabled() {
-    return collectionSettings.vectorConfig() != null
-        && collectionSettings.vectorConfig().vectorEnabled();
+  @SuppressWarnings("unchecked")
+  public CommandContext<TableSchemaObject> asTableContext() {
+    checkSchemaObjectType(TableSchemaObject.TYPE);
+    return (CommandContext<TableSchemaObject>) this;
   }
 
-  public IndexingProjector indexingProjector() {
-    return collectionSettings.indexingProjector();
+  @SuppressWarnings("unchecked")
+  public CommandContext<KeyspaceSchemaObject> asKeyspaceContext() {
+    checkSchemaObjectType(KeyspaceSchemaObject.TYPE);
+    return (CommandContext<KeyspaceSchemaObject>) this;
   }
+
+  @SuppressWarnings("unchecked")
+  public CommandContext<DatabaseSchemaObject> asDatabaseContext() {
+    checkSchemaObjectType(DatabaseSchemaObject.TYPE);
+    return (CommandContext<DatabaseSchemaObject>) this;
+  }
+
+  private void checkSchemaObjectType(SchemaObject.SchemaObjectType expectedType) {
+    Preconditions.checkArgument(
+        schemaObject().type == expectedType,
+        "SchemaObject type actual was %s expected was %s ",
+        schemaObject().type,
+        expectedType);
+  }
+
+  // TODO: why do we have these public ctors, and a static factor, and this is a record ??
```

Comment:
Need to remove unused; perhaps already done in later prs.
(as to why public -- Records can only have public members; and then factory methods for the usual reasons -- but codebase does overuse Records, some of them "just grew", others shouldn't have been Records in the first place).

Reply:
Removed unused. Not sure `CommandContext` should be Record, but not changing now.